src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,200 @@
 package core.basesyntax;
 
+import java.util.Iterator;
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        Node<T> last = tail;
+        Node<T> newNode = new Node<>(last, value, null);
+        tail = newNode;
+        if (last == null) {
+            head = newNode;
+        } else {
+            last.next = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Incorrect index " + index);
+        }
+        if (index == size) {
+            add(value);
+            return;
+        }
+        if (index == 0) {
+            addAtTheBeginning(value);
+            return;
+        }
+        Node<T> node = findNodeByIndex(index);
+        addBeforeNode(value, node);
     }
 
     @Override
     public void addAll(List<T> list) {
+        Iterator<T> iterator = list.iterator();
+        while (iterator.hasNext()) {
+            add(iterator.next());
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexBounds(index);
+        return findNodeByIndex(index).item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndexBounds(index);
+        Node<T> node = findNodeByIndex(index);
+        T oldItem = node.item;
+        node.item = value;
+        return oldItem;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexBounds(index);
+        if (index == size - 1) {
+            return removeLast();
+        }
+        if (index == 0) {
+            return removeFirst();
+        }
+        Node<T> nodeToRemove = findNodeByIndex(index);
+        unlink(nodeToRemove);
+        return nodeToRemove.item;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = head;
+        while (current != null) {
+            if (current.item == object || current.item != null
+                    && current.item.equals(object)) {
+                unlink(current);
+                return true;
+            }
+            current = current.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void addAtTheBeginning(T value) {
+        Node<T> first = head;
+        Node<T> newNode = new Node<>(null, value, first);
+        head = newNode;
+        if (first == null) {
+            tail = newNode;
+        } else {
+            first.prev = newNode;
+        }
```

Comment:
you can add logic from this block to the method `addBeforeNode(T value, Node<T> node)` and use only that method. That will optimize your code an program logic